File: leetcode/string/leetcode_402.java
package leetcode.string;

import java.util.Stack;

public class leetcode_402 {
    public static void main(String[] args) {
        leetcode_402 lc = new leetcode_402();
        System.out.println(lc.removeKdigits("1432219", 3));
    }

    public String removeKdigits(String num, int k) {
        int len = num.length();
        if (k == len)
            return "0";

        Stack<Character> stack = new Stack<>();
        for (char c : num.toCharArray()) {
            for (; k > 0 && !stack.isEmpty() && stack.peek() > c; k--) {
                stack.pop();
            }
            stack.push(c);
        }

        for (; k > 0; k--) {
            stack.pop();
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty())
            sb.append(stack.pop());
        sb.reverse();

        while (sb.length() > 1 && sb.charAt(0) == '0')
            sb.deleteCharAt(0);

        return sb.toString();
    }
}

File: string/LeetCode_402.kt
package string

import java.util.Stack

class LeetCode_402 {

    fun removeKdigits(num: String, k: Int): String {
        val n = num.length

        if (k == n) {
            return "0"
        }

        val stack = Stack<Char>()
        var k = k

        num.forEach {
            while (k > 0 && stack.isNotEmpty() && stack.peek() > it) {
                stack.pop()
                k--
            }

            stack.push(it)
        }

        repeat(k) {
            stack.pop()
        }

        val digitStringBuilder = StringBuilder()

        while (stack.isNotEmpty()) {
            digitStringBuilder.append(stack.pop())
        }

        digitStringBuilder.reverse()

        while (digitStringBuilder.length > 1 && digitStringBuilder.first() == '0') {
            digitStringBuilder.deleteAt(0)
        }

        return digitStringBuilder.toString()
    }
}

fun main() {
    // Input
    val num = "1432219"
    val k = 3
    // Output
    LeetCode_402().removeKdigits(num, k).run {
        println(this)
        require("1219" == this)
    }
}
